### PR TITLE
[Explicit Module Builds] Adopt dependency scanner Link Libraries support

### PR DESCRIPTION
**Pull request by artemcm**
- Adopt new libSwiftScan API to query link libraries for each module dependency
- Introduce '-explicit-auto-linking' to turn explicit dependency link library information into flags passed directly to the linker invocation

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci test

File: Sources/CSwiftScan/include/swiftscan_header.h [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #include <stdint.h>
 
 #define SWIFTSCAN_VERSION_MAJOR 0
-#define SWIFTSCAN_VERSION_MINOR 6
```
Comment (cachemeifyoucan) [inline Sources/CSwiftScan/include/swiftscan_header.h:21]: Can we check if we bump this, we are not missing any APIs that in the range of 7-9?

File: Sources/SwiftDriver/ExplicitModuleBuilds/ExplicitDependencyBuildPlanner.swift [hunk 2]
```diff
@@ -421,6 +421,29 @@ public typealias ExternalTargetModuleDetailsMap = [ModuleDependencyId: ExternalT
...
+    for (_, moduleInfo) in dependencyGraph.modules {
+      guard let moduleLinkLibraries = moduleInfo.linkLibraries else {
+        continue
+      }
+      for linkLibrary in moduleLinkLibraries {
+        allLinkLibraries.insert(linkLibrary)
+      }
+    }
+
+    for linkLibrary in allLinkLibraries {
```
Comment (cachemeifyoucan) [inline Sources/SwiftDriver/ExplicitModuleBuilds/ExplicitDependencyBuildPlanner.swift:435]: I don't know the corresponding scanner change. Does the library here guaranteed to be found by linker? Also does force-linking does anything in the link here in logics? Maybe considering just pass force-linked library with full path (so linker can't drop them) and stop emitting force-linking symbols.

**Review by cachemeifyoucan — COMMENTED** [review]
Looks good in general from swift-driver side. Need to double check the swift side.

File: Sources/SwiftDriver/ExplicitModuleBuilds/ExplicitDependencyBuildPlanner.swift [hunk 3]
```diff
@@ -421,6 +421,29 @@ public typealias ExternalTargetModuleDetailsMap = [ModuleDependencyId: ExternalT
     }
   }
 
+  public func getLinkLibraryLoadCommandFlags(_ commandLine: inout [Job.ArgTemplate]) throws {
+    var allLinkLibraries = Set<LinkLibraryInfo>()
```
[the remark this answers is not in the document]
Comment (artemcm) [reply Sources/SwiftDriver/ExplicitModuleBuilds/ExplicitDependencyBuildPlanner.swift]: Good catch. Thanks!

**Review by owenv — APPROVED** [review]

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci test Windows platform

**Pull request merged**